File: packages/xyd-opensdk-rust/src/rstype.ts
import type { TypeRef } from '@xyd-js/opensdk-core';

import { pascalCase } from './naming';

/**
 * Map an IR TypeRef to a Rust type expression (used for model fields, params
 * struct fields and method return types). Named refs render as their PascalCase
 * type name; arrays -> `Vec<T>`; maps -> `HashMap<String, T>`; a `binary`-format
 * scalar -> `Vec<u8>`; anything unclassified -> `serde_json::Value` (which any
 * JSON decodes into without error).
 */
export function rsType(ref: TypeRef | undefined): string {
  if (!ref) return 'serde_json::Value';
  switch (ref.kind) {
    case 'scalar':
      return rsScalar(ref.scalar, ref.format);
    case 'ref':
      return ref.name ? pascalCase(ref.name) : 'serde_json::Value';
    case 'array':
      return `Vec<${rsType(ref.items as TypeRef | undefined)}>`;
    case 'map':
      return `std::collections::HashMap<String, ${rsType(ref.values as TypeRef | undefined)}>`;
    default:
      return 'serde_json::Value';
  }
}

function rsScalar(scalar: string | undefined, format: string | undefined): string {
  if ((format || '').toLowerCase() === 'binary') return 'Vec<u8>';
  switch (scalar) {
    case 'string':
      return 'String';
    case 'integer':
      return format === 'int32' ? 'i32' : 'i64';
    case 'number':
      return format === 'float' ? 'f32' : 'f64';
    case 'boolean':
      return 'bool';
    default:
      return 'serde_json::Value';
  }
}

/** Whether a TypeRef is a `binary`-format scalar (rendered as `Vec<u8>`, sent as a file part). */
export function isBinaryRef(ref: TypeRef | undefined): boolean {
  return ref?.kind === 'scalar' && (ref.format || '').toLowerCase() === 'binary';
}

File: crates/xyd_opensdk_rust/src/rstype.rs
//! IR TypeRef → Rust type expression — port of rstype.ts.

use serde_json::Value;

use crate::naming::pascal_case;

/// Map a TypeRef (a JSON object) to a Rust type expression.
pub fn rs_type(ref_: Option<&Value>) -> String {
    let Some(r) = ref_ else {
        return "serde_json::Value".to_string();
    };
    match r.get("kind").and_then(|k| k.as_str()) {
        Some("scalar") => rs_scalar(
            r.get("scalar").and_then(|s| s.as_str()),
            r.get("format").and_then(|f| f.as_str()),
        ),
        Some("ref") => match r
            .get("name")
            .and_then(|n| n.as_str())
            .filter(|n| !n.is_empty())
        {
            Some(name) => pascal_case(name),
            None => "serde_json::Value".to_string(),
        },
        Some("array") => format!("Vec<{}>", rs_type(r.get("items"))),
        Some("map") => format!(
            "std::collections::HashMap<String, {}>",
            rs_type(r.get("values"))
        ),
        _ => "serde_json::Value".to_string(),
    }
}

fn rs_scalar(scalar: Option<&str>, format: Option<&str>) -> String {
    if format.unwrap_or("").to_lowercase() == "binary" {
        return "Vec<u8>".to_string();
    }
    match scalar {
        Some("string") => "String".to_string(),
        Some("integer") => {
            if format == Some("int32") {
                "i32".to_string()
            } else {
                "i64".to_string()
            }
        }
        Some("number") => {
            if format == Some("float") {
                "f32".to_string()
            } else {
                "f64".to_string()
            }
        }
        Some("boolean") => "bool".to_string(),
        _ => "serde_json::Value".to_string(),
    }
}

/// Whether a TypeRef is a `binary`-format scalar (rendered `Vec<u8>`).
pub fn is_binary_ref(ref_: Option<&Value>) -> bool {
    let Some(r) = ref_ else { return false };
    r.get("kind").and_then(|k| k.as_str()) == Some("scalar")
        && r.get("format")
            .and_then(|f| f.as_str())
            .unwrap_or("")
            .to_lowercase()
            == "binary"
}
